README.md
# RateThrottleClient

Rate limiting is for servers, rate throttling is for clients. This library implements a number of strategies for handling rate throttling on the client and a methodology for comparing performance of those clients. We don't just give you the code to rate throttle, we also give you the information to help you figure out the best strategy to rate throttle as well.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'rate_throttle_client'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install rate_throttle_client

## Usage

Wrap requests to an API endpoint using one of the provided rate throttling classes:

```ruby
throttle = RateThrottleClient::ExponentialIncreaseProportionalRemainingDecrease.new

response = throttle.call do
  Excon.get("https://api.example.com")
end
```

If the server returns a `429` status (the HTTP code indicating that a server side rate limit has been reached) then the request will be retried according to the classes' strategy.

## Expected return value from call

If you're not using Excon to build your API client, then you'll need to make sure the object returned to the block responds to `status` (returning the status code). To use `ExponentialIncreaseProportionalRemainingDecrease` it's expected that `headers["RateLimit-Remaining"].to_i` returns the number of available requests capacity.

### Config

```ruby
RateThrottleClient.config do |config|
  config.log_block = ->(info){ puts "I get called when rate throttling is triggered #{info.sleep_for} #{info.request}" }
  config.max_limit = 4500.to_f # Maximum number of requests available
  config.multiplier = 1.2 # When rate limiting happens, this is amount to the sleep value is increased by
end
```

## Strategies

This library has a few strategies you can choose between:

- RateThrottleClient::ExponentialBackoff
- RateThrottleClient::ExponentialIncreaseGradualDecrease
- RateThrottleClient::ExponentialIncreaseProportionalDecrease
- RateThrottleClient::ExponentialIncreaseProportionalRemainingDecrease

To choose, you need to understand what makes a "good" throttling strategy, and then you need some benchmarks.

## What Makes a Good Rate Throttle strategy?

- Minimize retry ratio: For example if every 50 successful requests, the client hits a rate limited request the ratio of retries is 1/50 or 2%. Why minimize this value? It takes CPU and Network resources to make requests that fail, if the client is making requests that are being limited, it's using resources that could be better spent somewhere else. The server also benefits as it spends less time dealing with rate limiting. (Tracked via Avg retry rate)
- Minimize standard deviation of request count across the system: If there are two clients and one client is throttling by sleeping for 100 seconds and the other is throttling for 1 second, the distribution of requests are not equitable. Ideally over time each client might go up or down, but both would see a median of 50 seconds of sleep time. Why? If processes in a system have a high variance, one process is starved for API resources. It then becomes difficult to balance or optimize otherworkloads. When a client is stuck waiting on the API, ideally it can perform other operations (for example in other threads). If one process is using 100% of CPU and slamming the API and other is using 1% of CPU and barely touching the API, it is difficult to balance the workloads. (Tracked via Stdev Request Count)
- Minimize sleep/wait time: Retry ratio can be improved artificially by choosing high sleep times. In the real world consumers don't want to wait longer than absolutely necessarry. While a client might be able to "work steal" while it is sleeping/waiting, there's not guarantee that's the case. Essentially assume that any amount of time spent sleeping over the minimum amount of time required is wasted. This value is calculateable, but that calculation requires complete information of the distributed system. (Tracked via Max sleep time)
  - At many available requests: It should be able to consume all available requests: If a server allows 100,000 requests in a day then a client should be capable of making 100,000 requests. If the rate limiting algorithm only allows it to make 100 requests it would have low retry ratio but high wait time.
  - At few available requests: If clients do not sleep enough their retry rate will be very high, if they sleep too much then they they are not are using available resources.
- Minimize time to respond to a change in available requests to either slow down or speed up rate throttling: A change can happen when clients are added or removed (for example if the number of servers/dynos are scaled up or down). It can also happen naturally if processing in a background worker or a web endpoint where the workload is cyclical. If there are few requests available and many become available, the rate throttle algorithm should adjust to match the new availability quickly. (Tracked via Time to clear workload)

The only strategy that handles all these scenarios well is currently: `RateThrottleClient::ExponentialIncreaseProportionalRemainingDecrease` against a [GCRA rate limit strategy, such as the one implemented by the Heroku API]().

## Pretty Chart

Here's a simulated run with the `ExponentialIncreaseProportionalRemainingDecrease` strategy with 10 clients (2 processes, 5 threads each) against a GCRA server for half an hour:

![](https://www.dropbox.com/s/2zda4v7jr07xy9d/2020-04-21-10-59-1587484779-399584000-RateThrottleClient%3A%3AExponentialIncreaseProportionalRemainingDecrease-chart.png?raw=1)

> A calculated optimal sleep time average for the system would be 8 seconds per client (arrival rate of new request capacity is 0.8 seconds, and there are 10 clients), this strategy oscilates around that value without needing complete information about the distributed system.

To generate a chart for all strategies run `rake bench`.

## Benchmarks

These benchmarks are generated by running `rake bench` against the simulated "[GCRA](https://brandur.org/rate-limiting)" rate limiting server.

**Lower values are better**

### RateThrottleClient::ExponentialBackoff results (duration: 30.0 minutes, multiplier: 1.2)

```
Avg retry rate:      80.41 %
Max sleep time:      46.72 seconds
Stdev Request Count: 147.84

Raw max_sleep_vals: [46.72, 46.72, 46.72, 46.72, 46.72, 46.50, 46.50, 46.50, 46.50, 46.50]
Raw retry_ratios: [0.79, 0.79, 0.81, 0.82, 0.80, 0.80, 0.79, 0.81, 0.82, 0.81]
Raw request_counts: [1317.00, 1314.00, 1015.00, 963.00, 1254.00, 1133.00, 1334.00, 1025.00, 1024.00, 1036.00]
```

```
Time to clear workload (4500 requests, starting_sleep: 1s):
74.33 seconds
```

### RateThrottleClient::ExponentialIncreaseGradualDecrease results (duration: 30.0 minutes, multiplier: 1.2)

```
Avg retry rate:      40.56 %
Max sleep time:      139.91 seconds
Stdev Request Count: 867.73

Raw max_sleep_vals: [110.25, 110.25, 110.25, 110.25, 110.25, 139.91, 139.91, 139.91, 139.91, 139.91]
Raw retry_ratios: [0.46, 0.37, 0.38, 0.37, 0.39, 0.40, 0.41, 0.35, 0.37, 0.57]
Raw request_counts: [48.00, 57.00, 56.00, 49.00, 282.00, 85.00, 83.00, 79.00, 2821.00, 37.00]
```

```
Time to clear workload (4500 requests, starting_sleep: 1s):
115.54 seconds
```

### RateThrottleClient::ExponentialIncreaseProportionalDecrease results (duration: 30.0 minutes, multiplier: 1.2)

```
Avg retry rate:      3.66 %
Max sleep time:      17.31 seconds
Stdev Request Count: 101.94

Raw max_sleep_vals: [17.31, 17.31, 17.31, 17.31, 17.31, 17.21, 17.21, 17.21, 17.21, 17.21]
Raw retry_ratios: [0.01, 0.07, 0.03, 0.05, 0.06, 0.01, 0.07, 0.01, 0.03, 0.03]
Raw request_counts: [343.00, 123.00, 223.00, 144.00, 128.00, 348.00, 116.00, 383.00, 194.00, 203.00]
```

```
Time to clear workload (4500 requests, starting_sleep: 1s):
551.10 seconds
```

### RateThrottleClient::ExponentialIncreaseProportionalRemainingDecrease results (duration: 30.0 minutes, multiplier: 1.2)

```
Avg retry rate:      3.07 %
Max sleep time:      17.32 seconds
Stdev Request Count: 78.44

Raw max_sleep_vals: [12.14, 12.14, 12.14, 12.14, 12.14, 17.32, 17.32, 17.32, 17.32, 17.32]
Raw retry_ratios: [0.03, 0.02, 0.01, 0.02, 0.03, 0.03, 0.02, 0.05, 0.04, 0.07]
Raw request_counts: [196.00, 269.00, 386.00, 302.00, 239.00, 197.00, 265.00, 150.00, 187.00, 118.00]
```

```
Time to clear workload (4500 requests, starting_sleep: 1s):
84.23 seconds
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/[USERNAME]/rate_throttle_client. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/[USERNAME]/rate_throttle_client/blob/master/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the RateThrottleClient project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/[USERNAME]/rate_throttle_client/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
## Master (Unreleased)

## 0.1.2

- Default rate throttle logging goes to stderr instead of stdout (https://github.com/zombocom/rate_throttle_client/pull/8)

## 0.1.1

- Supports Ruby 2.2
- Chart support

## 0.1.0

- First


lib/rate_throttle_client/demo.rb
require 'excon'
require 'pathname'
require 'fileutils'
require 'date'
require 'json'
require 'timecop'
require 'wait_for_it'
require 'enumerable/statistics'

require_relative 'chart.rb'

Thread.abort_on_exception = true

# A class for simulating or "demoing" a rate throttle client
#
# Example:
#
#   duration = 3600 # seconds in one hour
#   client = ExponentialIncreaseSleepAndRemainingDecrease.new
#   demo = RateThrottleDemo.new(client: client, stream_requests: true, duration: duration)
#   demo.call
#   demo.print_results
#     # => max_sleep_val: [59.05, 80.58, 80.58, 80.58, 56.18, 56.18, 56.18, 59.05, 70.05, 70.05, 70.05, 59.15, 59.15, 59.15, 59.15, 70.05, 70.05, 59.15, 56.18, 80.58, 80.58, 59.05, 59.05, 59.05, 56.18]
#     # => retry_ratio: [0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.02, 0.01, 0.01, 0.01, 0.00, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01]
#     # => request_count: [3321.00, 1551.00, 2167.00, 2197.00, 1628.00, 1709.00, 1484.00, 3512.00, 1722.00, 2816.00, 3182.00, 2137.00, 4398.00, 2154.00, 2418.00, 2868.00, 2492.00, 2982.00, 1731.00, 2278.00, 1988.00, 4221.00, 3160.00, 2927.00, 2635.00]
#
# Arguments:
#
# Thread count can be controlled via the `thread_count` arguement, or the THREAD_COUNT env var (default is below).
# Process count can be controlled via the `process_count` arguement, or the PROCESS_COUNT env var (default is below).
# total number of clients is thread_count * process_count.
#
# Time scale can be controlled via the `time_scale` arguement, or the TIME_SCALE env var (default is below).
# The time scale value will speed up the simulation, for example `TIME_SCALE=10` means a 60 second simulation
# will complete in 6 minutes.
#
# The simulation will stop after `duration:` seconds.
# Outputting request logs to stdout can be enabled/disabled by setting `stream_requests`.
# The other way to stop a simulation is to specify `remaining_stop_under` when this value is set, the simulation
# will stop when the "remaining" limit count from the server is under this value.
#
# The starting "remaining" limit count in the server can be set via passing in the `starting_limit`, default is 0
# requests.
#
# Outputs:
#
# - Writes log outputs to stdout if `stream_requests` is true
# - Writes aggregate metrics of each client to an intermediate json file every @json_duration seconds. This is then later used to
#   produce the data for `print_results`
# - Writes the last value the client slept for to a newline separated file every 1 (real time, not simulated) second. This is used to
#   generate the charts using the `chart.rb` script.
module RateThrottleClient
  class Demo
    MINUTE = 60
    THREAD_COUNT = ENV.fetch("THREAD_COUNT") { 5 }.to_i
    PROCESS_COUNT = ENV.fetch("PROCESS_COUNT") { 2 }.to_i
    DURATION=ENV.fetch("DURATION") { 30 }.to_i * MINUTE #
    TIME_SCALE = ENV.fetch("TIME_SCALE", 1).to_f
    RACKUP_FILE = Pathname.new(__dir__).join("servers/gcra/config.ru")

    attr_reader :log_dir, :rackup_file

    def initialize(client:,thread_count: THREAD_COUNT, process_count: PROCESS_COUNT, duration: DURATION, log_dir: nil, time_scale: TIME_SCALE, stream_requests: false, json_duration: 30, rackup_file: RACKUP_FILE, starting_limit: 0, remaining_stop_under: nil)
      @client = client
      @thread_count = thread_count
      @process_count = process_count
      @duration = duration
      @time_scale = time_scale.to_f
      @stream_requests = stream_requests
      @rackup_file = rackup_file
      @starting_limit = starting_limit
      @remaining_stop_under = remaining_stop_under

      if log_dir
        @log_dir = Pathname.new(log_dir)
      else
        @log_dir = Pathname.new(__dir__).join("../../logs/clients/#{Time.now.strftime('%Y-%m-%d-%H-%M-%s-%N')}-#{client.class}")
      end

      @mutex = Mutex.new
      @json_duration = 30 # seconds
      @port = UniquePort.call
      @threads = []
      @pids = []
      Timecop.scale(@time_scale)

      FileUtils.mkdir_p(@log_dir)
    end

    def print_results(io = STDOUT)
      result_hash = self.results
      io.puts
      io.puts "### #{@client.class} results (duration: #{@duration/60.0} minutes, multiplier: #{@client.multiplier})"
      io.puts
      io.puts "```"
      io.puts "Avg retry rate:      #{"%.2f" % (result_hash["retry_ratio"].mean * 100)} %"
      io.puts "Max sleep time:      #{"%.2f" % result_hash["max_sleep_val"].max} seconds"
      io.puts "Stdev Request Count: #{"%.2f" % result_hash["request_count"].stdev}"
      io.puts
      result_hash.each do |key, value|
        io.puts "Raw #{key}s: [#{ value.map {|x| "%.2f" % x}.join(", ")}]"
      end
      io.puts "```"
    end

    def chart(open_file)
      chart = RateThrottleClient::Chart.new(log_dir: @log_dir, name: @client.class.to_s.gsub("RateThrottleClient::", ""), time_scale: @time_scale)
      chart.call(open_file)
    end

    def results
      result_hash = {}

      @log_dir.entries.map do |entry|
        @log_dir.join(entry)
      end.select do |file|
        file.file? && file.extname == ".json"
      end.sort.map do |file|
        JSON.parse(file.read)
      end.each do |json|
        json.each_key do |key|
          result_hash[key] ||= []
          result_hash[key] << json[key]
        end
      end

      result_hash
    end

    def call
      WaitForIt.new("bundle exec puma #{@rackup_file.to_s} -p #{@port}", env: {"TIME_SCALE" => @time_scale.to_i.to_s, "STARTING_LIMIT" => @starting_limit.to_s}, wait_for: "Use Ctrl-C to stop") do |spawn|
        @process_count.times.each do
          boot_process
        end

        @pids.map { |pid| Process.wait(pid) }
      end
    end

    private def boot_process
      @pids << fork do
        run_threads
      end
    end

    private def run_threads
      @thread_count.times.each do
        @threads << Thread.new do
          run_client_single
        end
      end

      # Chart support, print out the sleep value in 1 second increments to a file
      Thread.new do
        loop do
          @threads.each do |thread|
            sleep_for = thread.thread_variable_get("last_sleep_value") || 0

            File.open(@log_dir.join("#{Process.pid}:#{thread.object_id}-chart-data.txt"), 'a') do |f|
              f.puts(sleep_for)
            end
          end
          sleep 1 # time gets adjusted via TIME_SCALE later in time.rb
        end
      end

      @threads.map(&:join)
    end

    class TimeIsUpError < StandardError; end

    private def run_client_single
      end_at_time = Time.now + @duration
      json_at_time = Time.now + @json_duration
      request_count = 0
      retry_count = 0

      monkey_patch_client_sleep
      loop do
        begin_time = Time.now
        break if begin_time > end_at_time

        if begin_time > json_at_time
          write_json_value(retry_count: retry_count, request_count: request_count, max_sleep_val: @client.max_sleep_val)
          json_at_time = begin_time + @json_duration
        end

        req = nil
        @client.call do
          request_count += 1

          req = make_request

          retry_count += 1 if req.status == 429

          stream_requests(req, retry_count: retry_count, request_count: request_count) if @stream_requests
          req
        end

        if @remaining_stop_under
          break if (req.headers["RateLimit-Remaining"].to_i <= @remaining_stop_under)
        end
      end
      stop_all_theads!

    rescue Excon::Error::Socket => e
      raise e
    rescue TimeIsUpError
      # Since the sleep time can be very high, we need a way to notify sleeping threads they can stop
      # When this exception is raised, do nothing and exit
    ensure
      write_json_value(retry_count: retry_count, request_count: request_count, max_sleep_val: @client.max_sleep_val)
    end

    private def monkey_patch_client_sleep
      @mutex.synchronize do
        if !@client.instance_variables.include?(:"@time_scale")
          def @client.sleep(val)
            @max_sleep_val = val if val > @max_sleep_val
            Thread.current.thread_variable_set("last_sleep_value", val)

            super val/@time_scale
          end

          def @client.max_sleep_val
            @max_sleep_val
          end

          def @client.last_sleep
            @last_sleep || 0
          end
        end

        @client.instance_variable_set(:"@time_scale", @time_scale)
        @client.instance_variable_set(:"@max_sleep_val", 0)
      end
    end

    private def make_request
      req = Excon.get("http://localhost:#{@port}")

      raise "Got unexpected reponse #{req.status}. #{req.inspect}" if req.status != 200 &&  req.status != 429
      req
    rescue Excon::Error::Timeout => e
      puts e.inspect
      puts "retrying"
      retry
    end

    private def stream_requests(request, retry_count:, request_count:)
      status_string = String.new
      status_string << "#{Process.pid}##{Thread.current.object_id}: "
      status_string << "status=#{request.status} "
      status_string << "remaining=#{request.headers["RateLimit-Remaining"]} "
      status_string << "retry_count=#{retry_count} "
      status_string << "request_count=#{request_count} "
      status_string << "max_sleep_val=#{ sprintf("%.2f", @client.max_sleep_val) } "

      puts status_string
    end

    # Even though all clients might have reached their `end_time` they might be stuck in a long `sleep`.
    # This method signals to any threads that might be stuck in a `sleep` to stop via an exception that we raise
    # and catch
    private def stop_all_theads!
      @threads.each do |t|
        next if @remaining_stop_under

        if t != Thread.current && t.backtrace_locations && t.backtrace_locations.first.label == "sleep"
          t.raise(TimeIsUpError)
        end
      end
    end

    private def write_json_value(retry_count:, request_count:, max_sleep_val:)
      results = {
        max_sleep_val: max_sleep_val,
        retry_ratio: retry_count / request_count.to_f,
        request_count: request_count
      }

      File.open(@log_dir.join("#{Process.pid}:#{Thread.current.object_id}.json"), 'w+') do |f|
        f.puts(results.to_json)
      end
    rescue TimeIsUpError
      retry
    end
  end

  require 'socket'

  module UniquePort
    def self.call
      TCPServer.open('127.0.0.1', 0) do |server|
        server.connect_address.ip_port
      end
    end
  end
end
